linalg/densemat.hpp

```diff
@@ -180,7 +192,7 @@ class DenseMatrix : public Matrix
 
    /// Compute y^t A x
    double InnerProduct(const Vector &x, const Vector &y) const
-   { return InnerProduct(x.HostRead(), y.HostRead()); }
+   { return InnerProduct(x.GetData(), y.GetData()); }
```

Comment:
Can you clarify why `GetData` is preferred over `HostRead` here?

Reply:
Here is my thought process on preferring `GetData` over `Host{Read,Write,ReadWrite}` for "small" vectors in this case (and in general where similar arguments apply as well, I think):

We consider all `DenseMatrix`es to be "small", so these vectors are also "small" and therefore their data should be always on host. Also, using `GetData` avoids two virtual calls to `HostRead` which for small sizes is definitely preferable.

I think calling this method with vector(s) in device memory should be considered an error, and in debug mode `GetData` will trigger a runtime error if the vector is valid on device. If a user really wants to use this method with device vectors then they can still do that by calling explicitly `InnerProduct(x.HostRead(), y.HostRead())` instead of `InnerProduct(x, y)`, or by explicitly moving `x` and `y` to host before the call.

If this does not sound convincing, we can definitely discuss it.

Reply:
I don't have a super strong opinion, but I would think `HostRead` is preferable to `GetData` for the reason that `HostRead` should be guaranteed to work, whereas `GetData` may cause a runtime failure.

While it makes sense that for the most part a `DenseMatrix` will be small (and therefore the associated `Vector`s will be on host), I'm not sure how widely known (or how well-documented) this convention is.

I hadn't noticed before that `Vector::HostRead` was virtual (I guess this was added for PETSc support in #1963).

I suppose essentially the choice boils down to choosing one of the two options:

1. It will be guaranteed to work and give the correct result, but there may be runtime overhead because of the virtual function call overhead.
2. It will be faster (avoiding the virtual calls), but not guaranteed to work (if the data is not valid on the host; runtime error in debug mode, crash(?) or silent failure(?) in optimized build).

Regardless of which option is chosen, the other behavior can still be obtained by manually calling the `double *` overload.

I guess which of the above two is preferable depends on just how significant the virtual function call overhead is in terms of application runtimes. @camierjs, did you look into these micro-benchmarks at some point?